main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -188,10 +188,11 @@ object GenExpression {
         mv.visitMethodInsn(INVOKESPECIAL, jvmName.toInternalName, JvmName.ConstructorMethod, MethodDescriptor.NothingToVoid.toDescriptor, false)
         // Capturing free args
         for ((arg, i) <- exps.zipWithIndex) {
-          val erasedArgType = BackendType.toErasedBackendType(arg.tpe)
+          val argType = BackendType.toBackendType(arg.tpe)
           mv.visitInsn(DUP)
           compileExpr(arg)
-          mv.visitFieldInsn(PUTFIELD, jvmName.toInternalName, s"clo$i", erasedArgType.toDescriptor)
+          BytecodeInstructions.castIfNotPrim(argType)
```

Comment:
When applying the closure argument, ideally we shouldn't have to cast after calling compileExpr, but the casting protocol of the backend still has some holes which forces us to cast here. Over time this should go away

Reply:
This means that in this PR, closure arguments are casted once upon capture, instead of casting them once at each call to the closure